keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,359 @@
+
+## Summary
+
+Kubeadm operator would like to enable declarative control of kubeadm workflows, automating the execution and the orchestration of such tasks across existing nodes in a cluster.
+
+## Motivation
+
+Kubeadm binary can execute actions only on the machine where it is running e.g. it is not possible to execute actions on other nodes, to copy files across nodes, etc.
+
+As a consequence, most of the kubeadm workflows, like e.g. [kubeadm upgrade](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/), consists of a complex sequence of tasks that should be manually executed/orchestrated across all the existing nodes in the cluster.
+
+Such a user experience is not ideal, potentially error-prone, and this approach can be considered also a limitation/blocker for implementing more complex workflows like e.g. rotating certificate authorities, change the settings of an existing cluster etc.
+
+This KEP aims to address such problems by applying the operator pattern to kubeadm workflows. 
+
+### Goals
+
+- To allow declarative control of kubeadm action that leads to "in place" mutations[1] of kubeadm generated artifacts. More specifically this initially targets includes kubeadm workflows:
+  - kubeadm upgrade
+  - certificate renewal
+  - change settings/configuration in an existing cluster (NEW) 
+  - certificate authority rotation (NEW)
+- To automate the orchestration of such mutations across the existing nodes in a cluster
+
+> [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight the difference between the kubeadm operator and other SCL projects like [Cluster API](https://cluster-api.sigs.k8s.io/), which instead assume node/machines immutable
+
+Considering the complexity of this topic, this document is expected to be subject to some iterations. The goal of the current iteration is to:
+
+- Get initial approval on Summary and Motivation paragraphs
+- To identify a semantic for defining “Actions” to be performed by the kubeadm-operator.
+- To define how the kubeadm-operator should manage kubeadm workflows.
+- To define how the users should interact with the kubeadm-operator, including also observability and error handling.
+- To define how the kubeadm-operator should be deployed in a Kubernetes cluster.
+
+### Non-Goals
+
+- To provide or manage any infrastructure elements like e.g. machines, load balancer, etc.
+- To manage/automate the kubeadm init/join workflows.
+- To provide or manage any component/configuration file/artifacts not generated by kubeadm, with the only exception of the kubelet/kubeadm/kubectl binaries which are considering in scope as required by the upgrade workflow.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes operator, I would like to be able to declaratively control upgrades in a systematic fashion.
```

Comment:
Not a native speaker, but I think `systematic` here looks weird. Automatic, unattended? Same for the other occurences.

Reply:
I think systematic makes sense if the intent is that there is some system to manage the configuration files that declare the workflows.